airflow/auth/managers/models/resource_details.py

```diff
@@ -0,0 +1,31 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from __future__ import annotations
+
+from dataclasses import dataclass
+
+
+@dataclass
+class ResourceDetails:
+    """
+    Represents the details of a resource.
+
+    All fields must be optional. These details can be used in authorization decision.
+    """
+
+    id: str | None = None
```

Comment:
I think we should have a dictionary describing the resource details. Rather than single detail, having a dictionary where we allow more flexible "identification" of the resource is likely better. For example for DAG we could keep both 'dag_id' but also "path" or even "file" but also other stuff ("user" etc.). Having a dictionary (and well defined structure of it "per resource typ" - seems more complex, but in fact it is esier. we are not tied with fields of ResourceDetails.

Reply:
Oh I see. But would not be the same then to create a ResourceDetails dataclass per resource type and allowing any dataclass as part of the `is_authorized` method? Somehow I prefer dataclass than dict :)

Reply:
I definitely dont like having multiple `is_authorized` methods. I really like the fact we have one entry point for authorization: `is_authorized`. It makes it cleaner and simpler to me (very subjective I know). What if we use dictionary? How could we use dictionary and have only one method `is_authorized`?

> Having a dictionary (and well defined structure of it "per resource typ" - seems more complex, but in fact it is esier. we are not tied with fields of ResourceDetails.

How are you thinking achieving this? Something like [schemadict](https://pypi.org/project/schemadict/) and we would have one schema per resource type? Performance wise I am not sure this is the right decision since `is_authorized` is called quite a lot

Reply:
How about having a single class/protocol that have multiple `is_authorized_dag(DagDetails)`,  `is_authorized_task(TaskDetails)` methods?

I am not super insisting on that but I just try to paint a picture of someone implementing and Auth Manager. If we have one method defined as it is currently, it will look more or less like that (unless I am mistaken and you can do it simpler):

```python
class MyAuthManager():
      def is_authorized(
        self,
        action: ResourceAction,
        resource_type: str,
        resource_details: ResourceDetails | None = None,
        user: BaseUser | None = None,
    ) -> bool:
       if resource_type = DAG:
           dag_details = cast(DagDetails, details)
           # here dag authorisation code
       elif resource_type = TASK:
           task_details = cast(TaskDetails, details)
           # here task authorisation code
      elif ...
```

Every single AuthManager will have to something like this I believe.

Whereas if you do multiple methods, then it will look quite a bit easier IMHO.

```python
class MyAuthManager():
      def is_authorized_dag(
        self,
        action: ResourceMethod,
        resource_details: DagDetails,
        user: BaseUser | None = None,
    ) -> bool:
           # here dag authorisation code

      def is_authorized_task(
        self,
        action: ResourceMethod,
        resource_details: TaskDetails,
        user: BaseUser | None = None,
    ) -> bool:
           # here tas authorisation code
```

If you ask me - as implementor, I DEFINITELY prefer the second case. I think the single method/pass resource type is forcing the implementor to do all the stuff that in the second case would be much more easily done in the Auth framework - there is no switching of type of resource information and no checking for None.

There are several other benefits of such solution IMHO:

1) when you add new resource type - you do not have to add another "if/else" - you just add another method. And you immediately know if the implementing class supports it - you can even check if it has the attribute defined and adapt in the framwork to serve the case that in the future some optional resources might not even be available if the method is not defined. In the first case, if you add a new resource type, you have to add yet another method "is_supported(resource_type)" and run in on the created instance of the AuthManager. In the second case it is enough to check just if the Class (not even instance) of the auth manager has appropriate method defined. 

2) the code for authentication of each type is smaller and more localized by default. In the first case with single is_auth_method, most implementation will have all the code in that method but as long as we have good programmers, they will - very quickly separate them to `_is_authorized_dag()`, `_is_authorized_task()` anyway - as soon as their if/elif/elif method will exceed 100 or so lines. So what eventually will happen - you will have those separate methods **anyway** - just without the benefit of your framework understanding, inspecting them .

3) This also immediately encourages the nice "DRY" pattern. if you want to have some common code that you will have in both _dag and _task (and other) - the first thing to do you will extract it to a separate method that will be used in both. There is basically no other way. and it's good. In the first case those "bad" programmers will keep the code in the `is_authorized` method which will make it veeeery long and difficult to test and reason about.  We have a few of those in our codebase... My favourite one (so favourite I even created an issue for that https://github.com/apache/airflow/issues/33249 ) is `backfill_job_runner` https://github.com/apache/airflow/blob/main/airflow/jobs/backfill_job_runner.py#L428 (but there are few others). By having a single `is_authorized` method - you are encouraging similar implementations I am afraid. By having them separate from day 0, you encourage more modular, focused and DRY implementations.

Reply:
I agree with **ALMOST** everything. At the end, I definitely agree with you we might end up having multiple private `_is_authorized_dag` ... There is one thing I dont really like with that though. We can certainly say that the resource attribute `id` is something common to all or the vast majority of resource types in Airflow. Example: when a user wants to modify a Variable `test-variable`, we definitely should call `is_authorized` with the variable ID `test-variable`. And so on for all resource types (there might be some exceptions). FAB auth manager might handle only the DAG ids but I dont think we should have this limitation in all auth managers. As a user/admin I'd like to be able to specify permissions by resource ID across all resource types. That would mean then having an `is_authorized` per resource type, which I think ... is massive.

So I have a proposal: why dont we add an optional parameter `resource_id` to the `is_authorized` method and we drop `resource_details`. This method would be used for any resource type with no additional details. For resources where additional details are needed when calling `is_authorized`, we would expose individual `is_authorized_` API with associated `resource_details` (basically, your proposal) without the resource ID since it is already provided as individual parameter.

WDYT?

Reply:
Challenge: Why massive?

How many resource types we are going to have ... Can we enumerate them now?

My guess is we will not have more than 10.

Reply:
Even if we have 10, does it make sense to have `is_authorized_variable`, `is_authorized_connection`, `is_authorized_connection`, `is_authorized_configuration` which most likely will have their own `ResourceDetails` which will reference only the ID?

Reply:
> Even if we have 10, does it make sense to have is_authorized_variable, is_authorized_connection, is_authorized_connection, is_authorized_configuration which most likely will have their own ResourceDetails which will reference only the ID?

Not at all. The methods for connection, configuration and everything else that just need id can be:
```
is_authorised_connection(connection_id: str)
is_authorised_configuration(configuration_id: str)
is_authorised_variable(variable_id: str)
```

I did not call for N resource types, just for N methods.

Reply:
I see, but then dont you think one generic method is enough? I guess it can depend on the auth manager bust most likely the implementation of all these methods will be strictly the same. You just need to pass this `resource_id` information to the backend service you are using to perform authorization

Reply:
But I just realized we would still need to test the resource type to format the request to the backend service. Ok, I am with you now :D

Reply:
Yeah. It's really the matter where you pass the resource type you ask for. There are two ways:

a) you pass the resource type as parameter of the method (or in some cases convention for resource type specification such as  "connection:aws_default_connection" - this is really what URI/URL is all about)
b) you make it inherently part of the "method" you call (so basically resource type is encoded in "which method you call")

My assumption is that in every place you want to check for resource is a python code where you already know what resource type you want to check. 
So it's perfectly fine by the python code to choose which method you call rather than specify parameter which resource type you want or building the right "URI" 

In our case we have both "caller" and "calee" a Python code (caller is our own "auth" framework, calee is the "authmanager" implementation). And we are in full control of the caller -> calee interface. Somehow it seems that having "one method" with "generic resource" adds overhead for no particular reason.

 It would force the "caller" to conver the "busines logic" it wants into a generic method os that the "calee" would again convert it back to the "use case". That seems like unnecessary overhead. We already have 'Python methods" as a transport layer for our intent - we do not need to map it to another generic construct.

It just seems more natural to  expres it 

* Caller wants to ask if the variable can be accessed by the user
* Callee is asked "is the variable accessible by the user"

Rather than:

* Caller wants to ask if the variable can be accessed by the user
* Caller converts this to "resource-type = variable" where "id=x"
* Callee converts the "resource-type" to variable check and "id" to "variable_id"
* Callee is asked "is the variable accessible by the user"

We can cut the two middle steps by explicitly asking for specific resource type via the choice of method call we make.

Reply:
My only worry is ... what if we add a new resource type in the future in Airflow (like `dataset` that has been introduced not a long time ago). We would need to create a `is_authorized_<new_resource_type>` in `base_auth_manager` but we cannot make it abstract because that would be breaking change. What should be the process here?

I am not saying this is a downside of using different methods because we would still have the same problem with one method (the unique method would not handle the new resource type) but I am just wondering what should be the process here in such situation

Reply:
No. It's better. We can make it part of Protocol (and anyone implementing Protocosl will be notified when it's missing) and we can check Auth Manager class to see if it actualy implements e each method. At leats we will know it does not and we can even hide accesse to paricular  UI if the Auth Manager effectively says "I hace no support for this auht tyoe.

Reply:
I have a second thought...

> How many resource types we are going to have ... Can we enumerate them now?

My understanding is the number of resource type will not change, hence from [permissions.py](https://github.com/apache/airflow/blob/main/airflow/security/permissions.py#L19), there are 40 resources types and as such, we should have 40 methods? This looks really massive.

Also, regarding FAB auth manager, even though we consider this auth manager as an exception, all `is_authorized_` methods are going to basically call one private `_is_authorized` method because this is the same implementation.

That makes me wonder whether other auth manager wont be in the same situation. We can easily imagine one auth manager with this pseudo code as implementation of `is_authorized`:

```
def is_authorized(
        self,
        action: ResourceMethod,
        resource_type: ResourceType,
        resource_details: ResourceDetails | None = None,
        user: BaseUser | None = None,
    ) -> bool:
    # Fetch the backend client of the service used tu make authorization decision
    client = ....

    return client.is_authorized(
        action=action.value,
        resource_type=resource_type.value
        extra=resource_details.to_dict(),
    )
```

This is obviously some simplified code and some logic is probably missing but the idea is to basically pass the information to the underlying service which will make the authorization decision.

Reply:
I am not insisting - if you will make a decision to make one "is_authorized" method because you think it will be better, I  have no big issue with it - but I would like to warn a bit from taking FAB as the "model" for resources. I think we want to break out from the model introduced by FAB because it was "bad". 

If you want to map the resources 1-1, yes you will have a lot of them. But I think we only have so many resources, because of FAB declarative method of defining the access via configuraiton. What we are planning to do is to implement AuthManager to define authentication as Code. it's very similar to what you'd compare Oozie and Airflow Workflows. If you want to convert oozie workflows to Airflow you'd end up with strange looking Airflow DAG, which you could much easier express in a Code in Airflow. 

The FAB "resource" definition has been traditionally very fine-grained because every permission in FAB was "declaratively" defined by assigning permission to role. This was really the only way to achieve it so if you wanted to decide if particular "sub-resource" is accessible or not in a fine grained way - you had no choice but to define fine-grained resources. And it was one of the FAB problems that it was too fine grained in definitiion. The only way you could give/not give access to particular view in FAB was that you combine "ACTION/METHOD" - ADD/DEL etc. with "RESOURCE" and either give or reject the access. 

For example if you wanted to decide to give/not give access to "DAG Code" you had to give access to RESOURCE_DAG_CODE. There was no other way. But. in our case we can define much more COARSE grained resources and decide in the Auth Manager implementation, whether it should be further fine-grained or not.

This is for example why we only have access "per DAG" as resource - because we had  no way to express it differeintly. You could not express a "group of DAGs" or "permission per tag" for that very reason because FAB model was "fixed" on resources and you could not express "Give permissions to all DAGS that have tag=x". 

But we can do soooo much better with AuthManagerAsCode.

We can programmatically decide individually what you do in each case - you do not have to have separate resource for every single action you do in them - you can make the decision internally in the implementaiton of the Auth manager.

For example instead of "DAG Code" resource have `auth_DAG` method. And this method can have both DagDetails and "accessing" field indicating which particular "part" of the DAG to access:

```python
def auth_dag(dag_details: DagDetails, task_details: TaskDetails. | None, accessing_action: DagAccessAction, method: Method):
    ....
```

Dag details:

```
DagDetails:
    id: 
    tags:
    queue/tenant:
    ....
```

Accessing Action: 

```
DagAccessAction(Enum):
     code,
     dependencies.
     run,
     task_instance,
     import_error.
```

```
MethodMethod):
    read,
    write,
    edit,
    delete
```

Now - our implementation might decide, based on whatever it wants -whether to allow or not specific action. It can be as simple as:

```
def auth_dag(dag_details: DagDetails, task_details: TaskDetails. | None, accessing_action: DagAccessAction, method: Method):
     if dag_details.queue == 'x':
         return True # Allow all DAG/TASK operations for DAG where queue is X
```

Or as complex as FAB -  mapping the above to the original "resources" and looking up the matrix of permission in the configuration and have access or not to particular tasks beloning to DAG based on whatever criteria.

Assumption here is that the permission set usually will be simple - a user can acces all dags that belong to the "queue" - if he belongs to admin group, "read only" if he is regular user, none if he is not there. 

I believe it is much simpler. I think if we do it this way, then implementor of each of the Auth Manager will make a decision on how complex, fine grained and configurable it is. If AWS Auth Manager want to **just** define 3 levels of users (Admin/ReadAll/None) and use "queue" to determine the group where those users are looked up - they will be able to do it. 



> My understanding is the number of resource type will not change, hence from [permissions.py](https://github.com/apache/airflow/blob/main/airflow/security/permissions.py#L19), there are 40 resources types and as such, we should have 40 methods? This looks really massive.

Yes if you map 1-1. But we should IMHO group access logically and leave the decision on "fine-graining" acces to the implementor.

Also a lot of those resources are not needed (user/admin/role goes away)

Here is my breakdown (very quick). I split the FAB resources into logical resources that could be used as "Auth manager" entrypoints - each entrypoint will have corresponding parameters determinining - if needed - not only the id of the resource but also action we want to do and method used.

User/Admin permissions  - not needed in the new model

```
RESOURCE_ACTION = "Permissions"
RESOURCE_ADMIN_MENU = "Admin"
RESOURCE_AIRFLOW = "Airflow"
RESOURCE_MY_PASSWORD = "My Password"
RESOURCE_MY_PROFILE = "My Profile"
RESOURCE_PASSWORD = "Passwords"
RESOURCE_PERMISSION = "Permission Views"  # Refers to a Perm <-> View mapping, not an MVC View.
RESOURCE_LOGIN = "Logins"
RESOURCE_ROLE = "Roles"
RESOURCE_USER = "Users"
RESOURCE_USER_STATS_CHART = "User Stats Chart"
```

Not needed because it can be implicit from other (connection/config etc).
```
RESOURCE_BROWSE = "Browse"
RESOURCE_RESOURCE = "View Menus"
```

We likely not need it:

```
RESOURCE_DOCS = "Documentation"
RESOURCE_DOCS_MENU = "Docs"
```



1. Pools:

```
RESOURCE_POOL = "Pools"
```

2. SLA:

```
RESOURCE_SLA_MISS = "SLA Misses"
```

3. Jobs:

```
RESOURCE_JOB = "Jobs"
```

4. Audit logs:

```
RESOURCE_AUDIT_LOG = "Audit Logs"
```

5. Configuration:

```
RESOURCE_CONFIG = "Configurations"
```

6. Connections:

```
RESOURCE_CONNECTION = "Connections"
```

7. Variables:

```
RESOURCE_VARIABLE = "Variables"
```

9. XComs:

```
RESOURCE_XCOM = "Xooms"
```

10. View (read-only state of the installation):

```
RESOURCE_PLUGIN = "Plugins"
RESOURCE_PROVIDER = "Providers"
RESOURCE_TRIGGER = "Triggers"
RESOURCE_WEBSITE = "Website"
```

11. Dags:

```
RESOURCE_DAG = "DAGs"
RESOURCE_DAG_CODE = "DAG Code"
RESOURCE_DAG_DEPENDENCIES = "DAG Dependencies"
RESOURCE_DAG_RUN = "DAG Runs"
RESOURCE_DAG_WARNING = "DAG Warnings"
RESOURCE_TASK_INSTANCE = "Task Instances"
RESOURCE_TASK_LOG = "Task Logs"
RESOURCE_TASK_RESCHEDULE = "Task Reschedules"
RESOURCE_IMPORT_ERROR = "ImportError"
```

12. Cluster activity:

```
RESOURCE_CLUSTER_ACTIVITY = "Cluster Activity"
```

13. Dataset

```
RESOURCE_DATASET = "Datasets"
```

I can meet and discuss more tomorrow if needed, I just wanted to make sure that we do not "stick" to the current - IMHO broken model of FAB but we come up with something much more reasonable.

Reply:
I see. I agree it makes more sense, I like the grouping of resources. I felt I followed blindly the way FAB defines resources but I definitely think, this new categorization is better. I think I'll go with that. I'll update the PR and see where it goes :)

As usual, thanks for the comments!

Reply:
Done. I really like it how it is now :) I agree it simplifies many things and the new "category" of resources it just great :D